crates/driver/src/domain/competition/auction.rs

```diff
@@ -76,12 +62,95 @@ impl Auction {
         self.id
     }
 
-    /// The orders for the auction. The orders are sorted such that those which
-    /// are more likely to be fulfilled come before less likely orders.
+    /// The orders for the auction.
     pub fn orders(&self) -> &[competition::Order] {
         &self.orders
     }
 
+    // Prioritize the orders such that those which are more likely to be fulfilled
+    // come before less likely orders. Filter out orders which the trader doesn't
+    // have enough balance to pay for.
+    //
+    // Prioritization is skipped during quoting. It's only used during competition.
+    pub async fn prioritize(mut self, eth: &Ethereum) -> Self {
+        // Sort orders so that most likely to be fulfilled come first.
+        self.orders.sort_by_key(|order| {
+            // Market orders are preferred over limit orders, as the expectation is that
+            // they should be immediately fulfillable. Liquidity orders come last, as they
+            // are the most niche and rarely used.
+            let class = match order.kind {
+                competition::order::Kind::Market => 2,
+                competition::order::Kind::Limit { .. } => 1,
+                competition::order::Kind::Liquidity => 0,
+            };
+            std::cmp::Reverse((
+                class,
+                // TODO No need to expose Tokens anymore, just use the auction
+                // If the orders are of the same kind, then sort by likelihood of fulfillment
+                // based on token prices.
+                order.likelihood(&self.tokens),
+            ))
+        });
+
+        // Fetch balances of each token for each trader.
+        let tokens_by_trader = self
+            .orders
+            .iter()
+            .flat_map(|order| {
+                [
+                    (order.trader(), order.sell.token),
+                    (order.trader(), order.buy.token),
+                ]
+            })
+            .unique()
+            .collect_vec();
+        let mut balances: HashMap<(order::Trader, eth::TokenAddress), eth::TokenAmount> =
+            join_all(tokens_by_trader.into_iter().map(|(trader, token)| {
```

Comment:
The same trader token pair will be fetched multiple times here. Maybe collect tokens_by_trader into a set instead of a vec.